tests/integration/crypto_tests.cpp

```diff
@@ -64,4 +64,169 @@ BOOST_FIXTURE_TEST_CASE( keccak_tests, tester ) try {
+} FC_LOG_AND_RETHROW()
+
+BOOST_FIXTURE_TEST_CASE( bn128_pair_tests, tester ) try {
+   create_accounts( { "test"_n } );
+   produce_block();
+
+   set_code( "eosio"_n,  contracts::crypto_primitives_test_wasm() );
+   set_abi(  "eosio"_n,  contracts::crypto_primitives_test_abi().data() );
+
+   produce_blocks();
+
+   push_action("eosio"_n, "pairtest"_n, "test"_n, mvo()
+      ("g1_a_x", "0f25929bcb43d5a57391564615c9e70a992b10eafa4db109709649cf48c50dd2")
+      ("g1_a_y", "16da2f5cb6be7a0aa72c440c53c9bbdfec6c36c7d515536431b3a865468acbba")
+      ("g2_a_x", "2e89718ad33c8bed92e210e81d1853435399a271913a6520736a4729cf0d51eb01a9e2ffa2e92599b68e44de5bcf354fa2642bd4f26b259daa6f7ce3ed57aeb3")
+      ("g2_a_y", "14a9a87b789a58af499b314e13c3d65bede56c07ea2d418d6874857b70763713178fb49a2d6cd347dc58973ff49613a20757d0fcc22079f9abd10c3baee24590")
+      ("g1_b_x", "1b9e027bd5cfc2cb5db82d4dc9677ac795ec500ecd47deee3b5da006d6d049b8")
+      ("g1_b_y", "11d7511c78158de484232fc68daf8a45cf217d1c2fae693ff5871e8752d73b21")
+      ("g2_b_x", "198e9393920d483a7260bfb731fb5d25f1aa493335a9e71297e485b7aef312c21800deef121f1e76426a00665e5c4479674322d4f75edadd46debd5cd992f6ed")
+      ("g2_b_y", "090689d0585ff075ec9e99ad690c3395bc4b313370b38ef355acdadcd122975b12c85ea5db8c6deb4aab71808dcb408fe3d1e7690c43d37b4ce6cc0166fa7daa")
+      ("expected", 0));
+} FC_LOG_AND_RETHROW()
+
+// Test G1 and G2 constructed from serialized points
+BOOST_FIXTURE_TEST_CASE( bn128_serialized_point_tests, tester ) try {
+   create_accounts( { "test"_n } );
+   produce_block();
+
+   set_code( "eosio"_n,  contracts::crypto_primitives_test_wasm() );
+   set_abi(  "eosio"_n,  contracts::crypto_primitives_test_abi().data() );
+
+   produce_blocks();
+
+   push_action("eosio"_n, "pairtest1"_n, "test"_n, mvo()
+      ("g1a", "0f25929bcb43d5a57391564615c9e70a992b10eafa4db109709649cf48c50dd216da2f5cb6be7a0aa72c440c53c9bbdfec6c36c7d515536431b3a865468acbba")
+      ("g2a", "2e89718ad33c8bed92e210e81d1853435399a271913a6520736a4729cf0d51eb01a9e2ffa2e92599b68e44de5bcf354fa2642bd4f26b259daa6f7ce3ed57aeb314a9a87b789a58af499b314e13c3d65bede56c07ea2d418d6874857b70763713178fb49a2d6cd347dc58973ff49613a20757d0fcc22079f9abd10c3baee24590")
+      ("g1b", "1b9e027bd5cfc2cb5db82d4dc9677ac795ec500ecd47deee3b5da006d6d049b811d7511c78158de484232fc68daf8a45cf217d1c2fae693ff5871e8752d73b21")
+      ("g2b", "198e9393920d483a7260bfb731fb5d25f1aa493335a9e71297e485b7aef312c21800deef121f1e76426a00665e5c4479674322d4f75edadd46debd5cd992f6ed090689d0585ff075ec9e99ad690c3395bc4b313370b38ef355acdadcd122975b12c85ea5db8c6deb4aab71808dcb408fe3d1e7690c43d37b4ce6cc0166fa7daa")
+      ("expected", 0));
+} FC_LOG_AND_RETHROW()
+
+BOOST_FIXTURE_TEST_CASE( modexp_tests, tester ) try {
+   create_accounts( { "test"_n } );
+   produce_block();
+
+   set_code( "eosio"_n,  contracts::crypto_primitives_test_wasm() );
+   set_abi(  "eosio"_n,  contracts::crypto_primitives_test_abi().data() );
+
+   produce_blocks();
+
```

Comment:
I suggest to also add additional test with smaller exponent and larger base number, same size as modulus (to emulate RSA crypto).

Reply:
Thanks. Will add more tests for this.